### PR TITLE
Codecov Report Publishing

### PR DESCRIPTION
**Pull request by trajinthan**
## Purpose

- Introducing `Codecov` code coverage report publishing to the `http` repository

## Goals

- Generating a code coverage report that is published via `Codecov` and triggered via GitHub actions of build and pull request.
- `Codecov` bot commenting on PRs

## Approach

- Custom execution file, java and ballerina class files, Sourcecode is provided to `jacoco` to generate an XML code coverage report.
- A new job is inserted to `Build` and `Pull request` GitHub actions that publishes the XML report to `Codecov`.
- These jobs trigger on merge to master or pull request and CodeCov generates a descriptive code coverage report.
- `Codecov` also adds comments to any PR made based on the code coverage changes between commits

### CODE REVIEW

**Comment by azinneera** [comment]
Acn we get the codecov to comment?

File: http-ballerina-tests/build.gradle [hunk 1]
```diff
@@ -26,6 +29,7 @@ def tomlVersion = project.version.replace("${snapshotVersion}", "")
 def ballerinaConfigFile = new File("$project.projectDir/Ballerina.toml")
 def ballerinaDependencyFile = new File("$project.projectDir/Dependencies.toml")
 def artifactCodeCoverageReport = file("$project.projectDir/target/cache/tests_cache/coverage/ballerina.exec")
+def artifactJar = file("$project.projectDir/build/target/extracted-distributions/jballerina-tools-zip/jballerina-tools-${project.ballerinaLangVersion}/repo/cache/ballerina/http/${tomlVersion}/java11/")
```
Comment (Kishanthan) [inline http-ballerina-tests/build.gradle:32]: Is this the correct path?

File: http-native/build.gradle [hunk 2]
```diff
@@ -129,3 +129,33 @@ compileJava {
         classpath = files()
     }
 }
+
+task extractJavaClassFiles(type: Copy) {
+    def tomlVersion = project.version.replace("-SNAPSHOT", "")
+    def jar = file("../http-ballerina-tests/build/target/extracted-distributions/jballerina-tools-zip/jballerina-tools-${project.ballerinaLangVersion}/repo/bala/ballerina/http/${tomlVersion}/platform/java11/http-native-${project.version}.jar")
```
Comment (azinneera) [inline http-native/build.gradle:135]: Can't we use the path from `build/lib` directory of http-native module

File: http-ballerina-tests/build.gradle (hunk 1 above)
Comment (azinneera) [reply http-ballerina-tests/build.gradle:32]: We can try copying the `cache` directory also here and assing the original location of the jar to `artifactJar` https://github.com/ballerina-platform/module-ballerina-http/blob/master/http-ballerina-tests/build.gradle#L97

File: http-native/build.gradle [hunk 3]
```diff
@@ -129,3 +129,33 @@ compileJava {
...
+task codeCoverageReportJava(type: JacocoReport) {
+    dependsOn(extractJavaClassFiles)
+    executionData file("${project.rootDir.absolutePath}/http-native/build/jacoco/test.exec")
+    additionalClassDirs files("${project.rootDir.absolutePath}/build/classes")
+
+    reports {
+        xml.enabled = true
+        html.enabled = true
+        xml.destination = new File("${buildDir}/reports/jacoco/report.xml")
+        html.destination = new File("${buildDir}/reports/jacoco/report.html")
```
Comment (azinneera) [inline http-native/build.gradle:155]: do we need the html report as well?

File: http-native/build.gradle (hunk 2 above)
Comment (trajinthan) [reply http-native/build.gradle:135]: yes, we can use. I will change the file path.

File: http-ballerina-tests/build.gradle (hunk 1 above)
Comment (trajinthan) [reply http-ballerina-tests/build.gradle:32]: I tried the approach but it didn't work. gave the same result as using the `http.jar` in `http-ballerina` without copying cache.

File: http-ballerina-tests/build.gradle [hunk 4]
```diff
@@ -212,3 +216,35 @@ publishing {
...
     }
 }
+
+task extractBallerinaClassFiles(type: Copy) {
+    def jars = fileTree(artifactJar).findAll {
+         it.name.endsWith(".jar")
+    }
+
+    jars.forEach { file ->
+        println(file.name)
```
Comment (Kishanthan) [inline http-ballerina-tests/build.gradle:226]: do we need this print ?

Comment (trajinthan) [reply http-ballerina-tests/build.gradle:226]: We don't need. I will remove it

Comment (trajinthan) [reply http-ballerina-tests/build.gradle]: commit ID [#29d784c](https://github.com/ballerina-platform/module-ballerina-http/pull/230/commits/29d784cbdbba0b9a63e304186dcb640a977da7e5)

**Review by azinneera — APPROVED** [review]

**Pull request merged**